Tests/AsyncHTTPClientTests/HTTPClientTestUtils.swift

```diff
@@ -1149,23 +1149,144 @@ struct CollectEverythingLogHandler: LogHandler {
     }
 }
 
+class StreamDelegate: HTTPClientResponseDelegate {
+    typealias Response = Void
+
+    enum State {
+        case idle
+        case waitingForBytes(EventLoopPromise<ByteBuffer?>)
+        case buffering(ByteBuffer, done: Bool)
+        case failed(Error)
+        case finished
+    }
+
+    let eventLoop: EventLoop
+    private var state: State = .idle
+
+    init(eventLoop: EventLoop) {
+        self.eventLoop = eventLoop
+    }
+
+    func next() -> EventLoopFuture<ByteBuffer?> {
+        if self.eventLoop.inEventLoop {
+            return self.next0()
+        } else {
+            return self.eventLoop.flatSubmit {
+                self.next0()
+            }
+        }
+    }
+
+    private func next0() -> EventLoopFuture<ByteBuffer?> {
+        switch self.state {
+        case .idle:
+            let promise = self.eventLoop.makePromise(of: ByteBuffer?.self)
+            self.state = .waitingForBytes(promise)
+            return promise.futureResult
+
+        case .buffering(let byteBuffer, done: false):
+            self.state = .idle
+            return self.eventLoop.makeSucceededFuture(byteBuffer)
+
+        case .buffering(let byteBuffer, done: true):
+            self.state = .finished
+            return self.eventLoop.makeSucceededFuture(byteBuffer)
+
+        case .waitingForBytes:
+            preconditionFailure("Don't call `.next` twice")
+
+        case .failed(let error):
+            self.state = .finished
+            return self.eventLoop.makeFailedFuture(error)
+
+        case .finished:
+            return self.eventLoop.makeSucceededFuture(nil)
+        }
+    }
+
+    // MARK: HTTPClientResponseDelegate
+
+    func didSendRequestHead(task: HTTPClient.Task<Response>, _ head: HTTPRequestHead) {
+        XCTAssert(self.eventLoop.inEventLoop)
+    }
+
+    func didSendRequestPart(task: HTTPClient.Task<Response>, _ part: IOData) {
+        XCTAssert(self.eventLoop.inEventLoop)
+    }
+
+    func didSendRequest(task: HTTPClient.Task<Response>) {
+        XCTAssert(self.eventLoop.inEventLoop)
```

Comment:
Can we remove this assertion? `inEventLoop` is validly allowed to return false negatives, so we shouldn't assert that it won't. We can add `preconditionInEventLoop` instead if you want the check.